### PR TITLE
433 : Improve performance of get plans

### PR DESCRIPTION
**Pull request by rehammuzzamil**
https://github.com/OpenSRP/opensrp-server-web/issues/433

### CODE REVIEW

File: src/main/java/org/opensrp/web/rest/PlanResource.java [hunk 1]
```diff
@@ -372,5 +401,30 @@ public boolean isReturnCount() {
...
+	private PlanSearchBean createPlanSearchBean(boolean isTemplateParam, Integer pageNumber, Integer pageSize, String orderByType,
+			String orderByFieldName, String planStatus, Map<String, String> useContextFilters) {
+		PlanSearchBean planSearchBean = new PlanSearchBean();
+		planSearchBean.setExperimental(isTemplateParam);
+		if(pageNumber != null) {
+			planSearchBean.setPageNumber(pageNumber);
+		}
+		if(pageSize != null) {
+			planSearchBean.setPageSize(pageSize);
+		}
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/PlanResource.java:414]: are the if statements required here

File: src/main/java/org/opensrp/web/rest/PlanResource.java [hunk 2]
```diff
@@ -372,5 +401,30 @@ public boolean isReturnCount() {
...
+		if (orderByFieldName != null) {
+			planSearchBean.setOrderByFieldName(PlanSearchBean.FieldName.valueOf(orderByFieldName));
+		}
+		if (planStatus != null) {
+			planSearchBean.setPlanStatus(PlanDefinition.PlanStatus.valueOf(planStatus));
+		}
+		if (useContextFilters != null) {
+			planSearchBean.setUseContexts(useContextFilters);
+		}
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/PlanResource.java:426]: are the if statements required here

File: src/main/java/org/opensrp/web/rest/PlanResource.java (hunk 1 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/PlanResource.java:414]: Setting fields in PlanSearchBean only when it's not null.

File: src/main/java/org/opensrp/web/rest/PlanResource.java (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/PlanResource.java:426]: Setting fields in PlanSearchBean only when it's not null.

Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/PlanResource.java]: Removed if statements to optimize complexity.

File: src/main/java/org/opensrp/web/rest/PlanResource.java (hunk 1 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/PlanResource.java]: Removed if statements to optimize complexity.

File: src/main/java/org/opensrp/web/rest/PlanResource.java (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/PlanResource.java]: Added if statements, where the **valueOf** method is being called to, get enum constant.

**Comment by rehammuzzamil** [comment]
This PR would consist of 
https://github.com/OpenSRP/opensrp-server-web/pull/596, 
https://github.com/OpenSRP/opensrp-server-web/pull/609, 
https://github.com/OpenSRP/opensrp-server-web/pull/597.
As discussed with @githengi , a single PR will be reviewed to avoid compilation errors as a single server-core tag is created for these PRs.

File: src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java [hunk 3]
```diff
@@ -63,8 +64,16 @@
 	@PostMapping(headers = { "Accept=multipart/form-data" })
 	public ResponseEntity<String> create(@RequestPart(required = false) MultipartFile file,
 			@RequestPart ProductCatalogue productCatalogue) {
+
+		final String DOWNLOAD_PHOTO_END_POINT = "/multimedia/media/";
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java:68]: can this be a static variable in this class

File: src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java [hunk 4]
```diff
@@ -95,7 +106,17 @@
 	public ResponseEntity<String> update(@PathVariable("id") Long uniqueId,
 			@RequestPart(required = false) MultipartFile file,
 			@RequestPart ProductCatalogue productCatalogue) {
+
+		final String DOWNLOAD_PHOTO_END_POINT = "/multimedia/media/";
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java:110]: can this be a static variable in this class

File: src/main/java/org/opensrp/web/rest/StockResource.java [hunk 5]
```diff
@@ -80,6 +80,14 @@
 
 	private static final String SAMPLE_CSV_FILE = "/importsummaryreport.csv";
 
+	public static final String PAGE_NUMBER = "pageNumber";
+
+	public static final String PAGE_SIZE = "pageSize";
+
+	public static final String ORDER_BY_TYPE = "orderByType";
+
+	public static final String ORDER_BY_FIELD_NAME = "orderByFieldName";
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/StockResource.java:89]: Define these centrally in `org.opensrp.web.Constants`

File: src/main/java/org/opensrp/web/rest/PlanResource.java [hunk 6]
```diff
@@ -80,6 +79,18 @@
...
+
+	public static final String PAGE_SIZE = "pageSize";
+
+	public static final String PLAN_STATUS = "planStatus";
+
+	public static final String USE_CONTEXT = "useContext";
+
+	public static final String ORDER_BY_TYPE = "orderByType";
+
+	public static final String ORDER_BY_FIELD_NAME = "orderByFieldName";
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/PlanResource.java:92]: Define these in `org.opensrp.web.Constants` except `PLAN_STATUS` and `USE_CONTEXT` which are used on this resource only

File: src/main/java/org/opensrp/web/rest/StockResource.java (hunk 5 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/StockResource.java]: Done

File: src/main/java/org/opensrp/web/rest/PlanResource.java (hunk 6 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/PlanResource.java]: Done

File: src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java (hunk 3 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java]: Done

File: src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java (hunk 4 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/ProductCatalogueResource.java]: Done

**Review by githengi — APPROVED** [review]

**Pull request merged**